src/platform_impl/windows/event_loop.rs

```diff
@@ -343,107 +350,115 @@ impl EventLoop {
-                    let handled = if let Some(callback) = self.msg_hook.as_deref_mut() {
-                        callback(&mut msg as *mut _ as *mut _)
-                    } else {
-                        false
-                    };
-                    if !handled {
-                        TranslateMessage(&msg);
-                        DispatchMessageW(&msg);
-                    }
-                }
+        let use_timer: bool;
+        if let (Some(handle), Some(timeout)) = (high_resolution_timer, timeout) {
+            let res = unsafe {
+                // Safety: handle can be Some only if we succeeded in creating high resolution
+                // timer. We properly clamped timeout so it can be used as argument
+                // to timer.
+                set_high_resolution_timer(handle, timeout)
+            };
+            if let Err(error_code) = res {
+                // We successfully got timer but failed to set it?
+                // Should be some bug in our code.
+                tracing::trace!("Failed to set high resolution timer: last error {}", error_code);
+                use_timer = false;
+            } else {
+                use_timer = true;
+            }
+        } else {
+            use_timer = false;
+        }
 
-                if let Err(payload) = runner.take_panic_error() {
-                    runner.reset_runner();
-                    panic::resume_unwind(payload);
-                }
-            },
+        unsafe {
+            // Either:
+            //  1. User wants to wait indefinely if timeout is not set.
+            //  2. We failed to get and set high resolution timer and we need something instead of
+            //     it.
+            let wait_duration_ms = timeout.map(dur2timeout).unwrap_or(INFINITE);
+
+            let mut raw_handles: [RawHandle; 1] = [ptr::null_mut()];
+            let (num_handles, handles_ptr) = if use_timer {
+                raw_handles[0] = high_resolution_timer.unwrap();
+                (1, raw_handles.as_ptr())
+            } else {
+                (0, ptr::null())
+            };
+
+            let result = MsgWaitForMultipleObjectsEx(
+                num_handles,
+                handles_ptr as *const _,
+                wait_duration_ms,
+                QS_ALLEVENTS,
+                MWMO_INPUTAVAILABLE,
+            );
+            if result == WAIT_FAILED {
+                // Well, nothing smart to do in such case.
+                // Treat it as spurious wake up.
+                tracing::warn!(
+                    "Failed to MsgWaitForMultipleObjectsEx: error code {}",
+                    GetLastError(),
+                );
+            }
         }
     }
 
     /// Dispatch all queued messages via `PeekMessageW`
     fn dispatch_peeked_messages(&mut self) {
         let runner = &self.window_target.runner_shared;
 
+        // Before we potentially exit, make sure to consistently emit an event for the wake up
+        runner.wakeup();
+
```

Comment:
Maybe this should be coupled with the implementation of `wait_for_messages`. 

Conceptually the intention was to have a sequence like:
1. notify the app that it's about to block/wait
2. _use OS API to block_
3. notify the app of a wake up
... then dispatch/handle all events that triggered the wake up

So long as `dispatch_peeked_messages` only gets called once immediately after `wait_for_messages` returns then notifying the wake up here is OK, but it feels to me like it would be more correct to notify about the wakeup immediately after `MsgWaitForMultipleObjectsEx` returns and avoid any risk of dispatching multiple wake ups if someone makes a change later that ends up calling `dispatch_peeked_messages` multiple times per wait.